### PR TITLE
Groupconv2d

### PR DESCRIPTION
**Pull request by zsdonghao**
#363

### CODE REVIEW

File: tensorlayer/layers/convolution.py [hunk 1]
```diff
@@ -1647,6 +1647,95 @@ def __init__(
...
+                name='weights',
+                shape=[filter_size[0], filter_size[1], channels / n_group, n_filter],
+                initializer=W_init,
+                dtype=D_TYPE,
+                trainable=True,
+                **W_init_args)
+            if b_init:
+                bi = tf.get_variable(
+                    name='biases', shape=[
+                        n_filter,
```
Comment (lgarithm) [inline tensorlayer/layers/convolution.py]: remove the comma after `n_filter` so we don't have to add a new line after it:
```
bi = tf.get_variable(name='biases', shape=[n_filter], initializer=b_init, dtype=D_TYPE, trainable=True, **b_init_args)
```

File: tensorlayer/layers/convolution.py [hunk 2]
```diff
@@ -1647,6 +1647,95 @@ def __init__(
...
+                    ], initializer=b_init, dtype=D_TYPE, trainable=True, **b_init_args)
+        if n_group == 1:
+            conv = groupConv(self.inputs, We)
+        else:
+            inputGroups = tf.split(axis=3, num_or_size_splits=n_group, value=self.inputs)
+            weightsGroups = tf.split(axis=3, num_or_size_splits=n_group, value=We)
+            convGroups = [groupConv(i, k) for i, k in zip(inputGroups, weightsGroups)]
+            conv = tf.concat(axis=3, values=convGroups)
+        if b_init is not None:
```
Comment (lgarithm) [inline tensorlayer/layers/convolution.py]:
```
if b_init:
```

File: tensorlayer/layers/convolution.py [hunk 3]
```diff
@@ -1647,6 +1647,95 @@ def __init__(
...
+            conv = tf.add(conv, bi, name='add')
+
+        self.outputs = act(conv)
+        self.all_layers = list(layer.all_layers)
+        self.all_params = list(layer.all_params)
+        self.all_drop = dict(layer.all_drop)
+        self.all_layers.extend([self.outputs])
+        if b_init is not None:
```
Comment (lgarithm) [inline tensorlayer/layers/convolution.py]:
```
if b_init:
```
I think we can simplify it as 

```
self.all_params.extend([We])
if b_init:
    self.all_params.extend([bi])
```

Comment (lgarithm) [reply tensorlayer/layers/convolution.py]: python list has an `append`method for adding a single element, so we can write
```
self.all_params.append(We)
```
but `extend` is used everywhere across the project, should we start using append for new code?
@luomai what do you think?

Comment (zsdonghao) [reply tensorlayer/layers/convolution.py]: all done

**Comment by lgarithm** [comment]
@zsdonghao Please use squash merge

**Pull request merged**